ql/src/java/org/apache/hadoop/hive/ql/optimizer/ParallelEdgeFixer.java

```diff
@@ -256,9 +257,20 @@ private static String extractColumnName(ExprNodeDesc expr) throws SemanticExcept
   public static Optional<Set<String>> colMappingInverseKeys(ReduceSinkOperator rs) {
     Map<String, String> ret = new HashMap<String, String>();
     Map<String, ExprNodeDesc> exprMap = rs.getColumnExprMap();
+    Set<String> neededColumns = new HashSet<String>();
     try {
       for (Entry<String, ExprNodeDesc> e : exprMap.entrySet()) {
-        ret.put(extractColumnName(e.getValue()), e.getKey());
+        String columnName = extractColumnName(e.getValue());
+        if (rs.getSchema().getColumnInfo(e.getKey()) == null) {
+          // ignore incorrectly mapped columns (if there's any) - but require its input to be present
+          neededColumns.add(columnName);
+        } else {
+          ret.put(columnName, e.getKey());
+        }
+      }
+      neededColumns.removeAll(ret.keySet());
+      if (!neededColumns.isEmpty()) {
+        throw new SemanticException("There is no way to compute: " + neededColumns);
```

Comment:
It would be useful to log the exception in the catch clause at least debug level.

Reply:
I've not logged these exceptions because a lot of them may happen during optimization - especially during swo tries uncover new possibilities.

Its already a challenging to make sense out from only looking at the swo logs.

But some logging will still happen around [here](https://github.com/apache/hive/blob/fb25765b831cc9269d104b5825050a7cbee0d8b6/ql/src/java/org/apache/hadoop/hive/ql/optimizer/SharedWorkOptimizer.java#L1852)  why the optimization may not kick in.